types/math/dec.go

```diff
@@ -0,0 +1,126 @@
+package math
+
+import (
+	"github.com/cockroachdb/apd/v2"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type Dec struct {
+	dec apd.Decimal
+}
+
+var dec128Context = apd.Context{
+	Precision:   34,
+	MaxExponent: apd.MaxExponent,
+	MinExponent: apd.MinExponent,
+	Traps:       apd.DefaultTraps,
+}
+
+func NewDecFromString(s string) (Dec, error) {
+	d, _, err := apd.NewFromString(s)
+	if err != nil {
+		return Dec{}, err
+	}
+	return Dec{*d}, nil
+}
+
+func NewDecFromInt64(x int64) Dec {
+	var res Dec
+	res.dec.SetInt64(x)
+	return res
+}
+
+func (x Dec) Add(y Dec) (Dec, error) {
+	var z Dec
+	_, err := apd.BaseContext.Add(&z.dec, &x.dec, &y.dec)
```

Comment:
let's add `SafeAdd` or check here (in `Add`) that both `x` and `y` are not negative.

Reply:
What's wrong with either `x` or `y` being negative? There's tests in here that include usage of negative `Dec` values, I'm not sure why we would want to enforce at the library level that this function cannot work with negative values.

Can you explain the use case for a `SafeAdd` ? `SafeSub` makes more sense to me, as verifying that balance is non-zero is in fact something we often have to do. I can't really think of a use case for `SafeAdd` though.

Reply:
I think this relates to https://github.com/regen-network/regen-ledger/pull/307/files#r678209627

Reply:
> What's wrong with either x or y being negative?

@clevinson  we want to have "Safe" operations - operations which are assured to not return negative values. `2 + (-3)` is negative.